File: vane-core/src/main/java/org/oddlama/vane/core/config/ConfigIntListField.java
package org.oddlama.vane.core.config;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.Function;
import org.apache.commons.lang3.ArrayUtils;
import org.bukkit.configuration.file.YamlConfiguration;
import org.oddlama.vane.annotation.config.ConfigIntList;
import org.oddlama.vane.core.YamlLoadException;

public class ConfigIntListField extends ConfigField<List<Integer>> {

    public ConfigIntList annotation;

    public ConfigIntListField(Object owner, Field field, Function<String, String> mapName, ConfigIntList annotation) {
        super(owner, field, mapName, "int list", annotation.desc());
        this.annotation = annotation;
    }

    private void appendIntListDefinition(StringBuilder builder, String indent, String prefix, List<Integer> def) {
        appendListDefinition(builder, indent, prefix, def, (b, i) -> b.append(i));
    }

    @Override
    public List<Integer> def() {
        final var override = overriddenDef();
        if (override != null) {
            return override;
        } else {
            return Arrays.asList(ArrayUtils.toObject(annotation.def()));
        }
    }

    @Override
    public boolean metrics() {
        final var override = overriddenMetrics();
        if (override != null) {
            return override;
        } else {
            return annotation.metrics();
        }
    }

    @Override
    public void generateYaml(StringBuilder builder, String indent, YamlConfiguration existingCompatibleConfig) {
        appendDescription(builder, indent);
        appendValueRange(builder, indent, annotation.min(), annotation.max(), Integer.MIN_VALUE, Integer.MAX_VALUE);

        // Default
        builder.append(indent);
        builder.append("# Default:\n");
        appendIntListDefinition(builder, indent, "# ", def());

        // Definition
        builder.append(indent);
        builder.append(basename());
        builder.append(":\n");
        final var def = existingCompatibleConfig != null && existingCompatibleConfig.contains(yamlPath())
            ? loadFromYaml(existingCompatibleConfig)
            : def();
        appendIntListDefinition(builder, indent, "", def);
    }

    @Override
    public void checkLoadable(YamlConfiguration yaml) throws YamlLoadException {
        checkYamlPath(yaml);

        if (!yaml.isList(yamlPath())) {
            throw new YamlLoadException("Invalid type for yaml path '" + yamlPath() + "', expected list");
        }

        for (var obj : yaml.getList(yamlPath())) {
            if (!(obj instanceof Number)) {
                throw new YamlLoadException("Invalid type for yaml path '" + yamlPath() + "', expected int");
            }

            var val = yaml.getInt(yamlPath());
            if (annotation.min() != Integer.MIN_VALUE && val < annotation.min()) {
                throw new YamlLoadException(
                    "Configuration '" + yamlPath() + "' has an invalid value: Value must be >= " + annotation.min()
                );
            }
            if (annotation.max() != Integer.MAX_VALUE && val > annotation.max()) {
                throw new YamlLoadException(
                    "Configuration '" + yamlPath() + "' has an invalid value: Value must be <= " + annotation.max()
                );
            }
        }
    }

    public List<Integer> loadFromYaml(YamlConfiguration yaml) {
        final var list = new ArrayList<Integer>();
        for (var obj : yaml.getList(yamlPath())) {
            list.add(((Number) obj).intValue());
        }
        return list;
    }

    public void load(YamlConfiguration yaml) {
        try {
            field.set(owner, loadFromYaml(yaml));
        } catch (IllegalAccessException e) {
            throw new RuntimeException("Invalid field access on '" + field.getName() + "'. This is a bug.");
        }
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/config/ConfigIntListField.kt
package org.oddlama.vane.core.config

import org.bukkit.configuration.file.YamlConfiguration
import org.oddlama.vane.annotation.config.ConfigIntList
import org.oddlama.vane.core.YamlLoadException
import java.lang.reflect.Field

/**
 * Config field handler for integer list values.
 *
 * @param owner object containing the reflected config field.
 * @param field reflected config field.
 * @param mapName maps Java field names to YAML paths.
 * @param annotation source annotation metadata.
 */
class ConfigIntListField(
    owner: Any?,
    field: Field,
    mapName: (String?) -> String?,
    /** Annotation metadata for this field. */
    var annotation: ConfigIntList
) : ConfigField<MutableList<Int?>?>(owner, field, mapName, "int list", annotation.desc) {

    /** Appends an integer list definition block. */
    private fun appendIntListDefinition(
        builder: StringBuilder?,
        indent: String?,
        prefix: String?,
        def: MutableList<Int?>
    ) {
        appendListDefinition<Int?>(builder, indent, prefix, def) { b, i -> b!!.append(i!!) }
    }

    /** Returns the default value for this config field. */
    override fun def(): MutableList<Int?> = overriddenDef() ?: annotation.def.map { it as Int? }.toMutableList()

    /** Returns whether metrics collection is enabled for this field. */
    override fun metrics(): Boolean = overriddenMetrics() ?: annotation.metrics

    /** Generates YAML for this field. */
    override fun generateYaml(builder: StringBuilder, indent: String, existingCompatibleConfig: YamlConfiguration?) {
        appendDescription(builder, indent)
        appendValueRange(builder, indent, annotation.min, annotation.max, Int.MIN_VALUE, Int.MAX_VALUE)
        builder.append("$indent# Default:\n")
        appendIntListDefinition(builder, indent, "# ", def())
        builder.append("$indent${basename()}:\n")
        val def = if (existingCompatibleConfig != null && existingCompatibleConfig.contains(yamlPath()))
            loadFromYaml(existingCompatibleConfig) else def()
        appendIntListDefinition(builder, indent, "", def)
    }

    /** Validates that this field is loadable from YAML. */
    @Throws(YamlLoadException::class)
    override fun checkLoadable(yaml: YamlConfiguration) {
        checkYamlPath(yaml)
        if (!yaml.isList(yamlPath()))
            throw YamlLoadException("Invalid type for yaml path '${yamlPath()}', expected list")
        for (obj in yaml.getList(yamlPath())!!) {
            if (obj !is Number)
                throw YamlLoadException("Invalid type for yaml path '${yamlPath()}', expected int")
            validateIntRange(yamlPath(), obj.toInt(), annotation.min, annotation.max)
        }
    }

    /** Loads this field value from YAML. */
    fun loadFromYaml(yaml: YamlConfiguration): MutableList<Int?> =
        yaml.getList(yamlPath())!!.map { (it as Number).toInt() as Int? }.toMutableList()

    /** Writes the loaded value into the reflected field. */
    override fun load(yaml: YamlConfiguration) {
        try {
            field.set(owner, loadFromYaml(yaml))
        } catch (_: IllegalAccessException) {
            throw RuntimeException("Invalid field access on '${field.name}'. This is a bug.")
        }
    }
}